astropy/cosmology/parameter/_core.py

```diff
@@ -1,14 +1,68 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 
+from __future__ import annotations
+
 import copy
+from dataclasses import dataclass, field, fields, replace
+from typing import Any, Sequence
 
 import astropy.units as u
+from astropy.utils.compat import PYTHON_LT_3_10
 
-from ._converter import _REGISTRY_FVALIDATORS, _register_validator
+from ._converter import _REGISTRY_FVALIDATORS, FValidateCallable, _register_validator
 
 __all__ = []
 
 
+if not PYTHON_LT_3_10:
+    from dataclasses import KW_ONLY
+else:
+    KW_ONLY = Any
+
+
+@dataclass(frozen=True)
+class _UnitField:
+    def __get__(
+        self, obj: Parameter | None, objcls: type[Parameter] | None
+    ) -> u.Unit | None:
```

Comment:
> Are you sure that `u.Unit` is the correct type here and not `u.UnitBase`

The `__setattr__` always wraps the value as `u.Unit(value)`, so `u.Unit` should be the correct type here.
(a strange emphasis on should, since `u.Unit` is strange).

> what would be the downside of replacing None with `u.dimensionless_unscaled`

If the value doesn't have a unit, as in it isn't a `Quantity`, which is distinct from `u.dimensionless_unscaled`.

> The upside would be that the return type wouldn't be a union type.

Agreed, but see above. As a note, part of my plan for Cosmology is to make more fields, e.g. `Om0`, have units, even if the unit is `u.dimensionless_unscaled`.

Reply:
`u.Unit` is indeed strange:
```python
>>> from astropy import units as u
>>> isinstance(u.m, u.Unit)
False
>>> type(u.m).__mro__
(<class 'astropy.units.core.IrreducibleUnit'>, <class 'astropy.units.core.NamedUnit'>, <class 'astropy.units.core.UnitBase'>, <class 'object'>)
```

I would expect that all cosmology parameters could be represented as `Quantity` instances, even if some of them would require `u.dimensionless_unscaled` as the unit, but if that is not the case currently then it is indeed better to leave it as it is for now.